src/com/cft/contactmerge/FirstAndLastNamesMatchLogic.java

```diff
@@ -28,13 +73,145 @@ private static AnswerType getMatchingResultFromMatchingMethodsList(String contac
+                firstNameIsInitialOf2ndFirstName = true;
+            }
+        } else if (isNameInitial(contactTwoFirstNameWithoutPunctuations)){
+            if (contactOneFirstNameWithoutPunctuations.startsWith(contactTwoFirstNameWithoutPunctuations)){
+                firstNameIsInitialOf2ndFirstName = true;
+            }
+        }
+
+        // Last name is an initial
+        if (isNameInitial(contactOneLastNameWithoutPunctuations)){
+            // Initial matches the last name
+            if (contactTwoLastNameWithoutPunctuations.startsWith(contactOneLastNameWithoutPunctuations)){
+                lastNameIsInitialOf2ndLastName = true;
+            }
+        } else if (isNameInitial(contactTwoLastNameWithoutPunctuations)){
+            if (contactOneLastNameWithoutPunctuations.startsWith(contactTwoLastNameWithoutPunctuations)){
+                lastNameIsInitialOf2ndLastName = true;
+            }
+        }
+
+        // If names were not an initial, just check if they match.
+        if (!firstNameIsInitialOf2ndFirstName){
+            if (contactOneFirstNameWithoutPunctuations.equals(contactTwoFirstNameWithoutPunctuations)){
+                firstNamesMatchFound = true;
+            }
+        }
+
+        // If names were not an initial, just check if they match.
+        if (!lastNameIsInitialOf2ndLastName){
+            if (contactOneLastNameWithoutPunctuations.equals(contactTwoLastNameWithoutPunctuations)){
+                lastNamesMatchFound = true;
+            }
+        }
+
+        // Return matching logic
+        if ((firstNameIsInitialOf2ndFirstName && lastNameIsInitialOf2ndLastName) ||
+                (firstNameIsInitialOf2ndFirstName && lastNamesMatchFound) ||
+                (firstNamesMatchFound && lastNameIsInitialOf2ndLastName) ||
+                (firstNamesMatchFound && lastNamesMatchFound)){
+            return true;
+        }
+        return false;
+    }
+
+    /*******************
+     * Support Methods *
+     *******************/
+
+    private static boolean isNameInGroupNamesMatch(ArrayList<String> namesOne, ArrayList<String> namesTwo){
+        for (String name: namesOne){
+            if (namesTwo.contains(name)){
+                return true;
+            }
+        }
+        return false;
+    }
+
+    private static boolean isNameInitial(String name){
+        return name.length() == 1;
+    }
+
+    private static ArrayList<String> seperateHypenNames(String name){
+        return new ArrayList<>(Arrays.asList(name.split("-",-1)));
+    }
+
+    private static String replacePunctuationsInNameWithWhiteSpace(String name){
+        String nameWithoutPunctuations = name;
+
+        for (String punctuation: namePunctuations){
+            nameWithoutPunctuations = nameWithoutPunctuations.replace(punctuation, " ");
+        }
+        return nameWithoutPunctuations;
+    }
+
+    private static String removePunctuationsInName(String name){
```

Comment:
The code in this method is identical to the code in replacePunctuationsInNameWithWhiteSpace(). I don't think we need this version as the test with whitespace removed is the one we want to use all the time anyways.

Reply:
Ah... I see the difference now. One you replace with a space and another with an empty string. Do we really need both? If yes, still should refactor to reduce duplicate code. One way would be to create another method that accepts the replacement character and then each of the existing methods would call with the appropriate replacement character.

Reply:
I agree with passing the replacement character as an argument to remove duplicate code. When I refactor the code, I may not need both. I was thinking in terms of handling separate cases. But if we are set on a generic matching logic, I probably will only be replacing with white spaces. I will have to see....